corehq/apps/sso/models.py

```diff
@@ -418,6 +426,71 @@ def get_required_identity_provider(cls, username):
             return idp
         return None
 
+    def get_all_members_of_the_idp(self):
+        import msal
+        authority_base_url = "https://login.microsoftonline.com/"
+        authority = f"{authority_base_url}{self.api_host}"
+
+        config = {
+            "authority": authority,
+            "client_id": self.api_id,
+            "scope": ["https://graph.microsoft.com/.default"],
+            "secret": self.api_secret,
+            "endpoint": "https://graph.microsoft.com/v1.0/users"
+        }
+
+        try:
+            # Create a preferably long-lived app instance which maintains a token cache.
+            app = msal.ConfidentialClientApplication(
+                config["client_id"], authority=config["authority"],
+                client_credential=config["secret"],
+            )
+            # looks up a token from cache
+            result = app.acquire_token_silent(config["scope"], account=None)
+            if not result:
+                result = app.acquire_token_for_client(scopes=config["scope"])
+            if "access_token" in result:
+                # Calling graph using the access token
+                response = requests.get(
+                    config["endpoint"],
+                    headers={'Authorization': 'Bearer ' + result['access_token']},
+                )
+                response.raise_for_status()  # Raises an error for bad status
+                graph_data = response.json()
+
+                user_principal_names = [user["userPrincipalName"] for user in graph_data["value"]]
+                return user_principal_names
+            else:
+                error_message = f"Error acquiring Microsoft Graph API token: {result.get('error')}, Description: "
+                f"{result.get('error_description')}, Correlation ID: {result.get('correlation_id')}"
+                raise Exception(error_message)
+        except Exception as e:
+            notify_exception(None, f"Failed to get members of the IdP. {str(e)}")
+
+            context = get_graph_api_connection_issue_email_context(self, str(e))
+            if not context["to"]:
+                notify_exception(None, f"no admin email addresses for IdP: {self}")
+            try:
+                for send_to in context["to"]:
+                    send_html_email_async.delay(
+                        context["subject"],
+                        send_to,
+                        context["html"],
+                        text_content=context["plaintext"],
+                        email_from=context["from"],
+                        bcc=context["bcc"],
+                    )
+                    log.info(
+                        "Sent Microsoft Graph API connection issue email "
+                        "for %(idp_name)s to %(send_to)s." % {
+                            "idp_name": self.name,
+                            "send_to": send_to,
+                        }
+                    )
+            except Exception as exc:
```

Comment:
Is there a specific exception you expect to be raised from this? It looks like fairly straightforward code that triggers celery tasks.